### PR TITLE
feat: helm chart deployment with multi-shards

### PR DESCRIPTION
**Pull request by numb3r3**
# Deploy QPLite Using Helm

## Prerequisite
- Kubernetes Cluster Setup
- Clone [this](https://github.com/jina-ai/pqlite.git) git repo

## Step 1: Clone Helm Repository

```bash
$ git clone https://github.com/jina-ai/pqlite.git \
  && cd pqlite
```

## Step 2: Adding a Bitnami Chart dependency

Chart dependencies are used to install other charts’ resources that a Helm chart may depend on.

In this chart, we are using bitnami/common as a plugin so we to need add this as a dependency.

When downloading a dependency for the first time, you should use the helm dependency update command.

```bash
$ helm dependency update ./pqlite-helm
```

## Step 3: Create and Apply Persistent Storage Volume

The data in your Pqlite indexer need to persist across pod restarts.

To achieve this, create a PersistentVolume resource in a YAML file. This example `pqlite-storage.yaml` uses the following configuration:
```YAML
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pqlite-pv-volume
  labels:
    type: local
    app: pqlite
spec:
  storageClassName: manual
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteMany
  hostPath:
    path: "/tmp/pqlite"
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pqlite-pvc-claim
  labels:
    app: pqlite
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 5Gi
```

Then apply the configuration with `kubectl apply`:

```bash
$ kubectl apply -f pqlite-storage.yaml
```

## Step 4: Install Helm Chart

Install the helm chart with the `helm install` command. Add `--set` flags to the command to connect the installation to the PVC you created and enable volume permissions:

```bash
$ helm install pqlite-test \
    --set dimension=256 \
    --set shardCount=3 \
    --set persistence.existingClaim=pqlite-pvc-claim \
    --set persistence.enabled=true \
    ./pqlite-helm
```

## Step 5: Use in a Jina Flow

Expose port `4567` to access the service via `kubectl port-forward ...`

```bash
$ kubectl port-forward service/indexer-head 4567:8081
```

This example shows how to use in a Jina Flow

```python
import numpy as np
from jina import DocumentArray, Flow

PROTOCOL = 'grpc'
EXTERNAL_HEAD_PORT_IN = 4567

f = Flow(protocol=PROTOCOL).add(
    name='indexer',
    external=True,
    host='localhost',
    port_in=EXTERNAL_HEAD_PORT_IN,
)

with f:

    docs = DocumentArray.empty(50)
    docs.embeddings = np.random.random((50, 512)).astype(np.float32)

    result = f.post('/index', inputs=docs, request_size=1, return_results=True)
    result = f.post('/search', inputs=docs[:5], return_results=True)
```

## Conferences

- [https://phoenixnap.com/kb/postgresql-kubernetes](https://phoenixnap.com/kb/postgresql-kubernetes)

### CODE REVIEW

**Review by JoanFM — COMMENTED** [review]
I think this helm chart should not include the `gateway`

**Comment by numb3r3** [comment]
Yes, the `gateway` is optional. The initial motivation for this PR is to support **PQLite-As-A-Service**, and then `gateway` is added to expose the HTTP (or GRPC) service.

File: pqlite-helm/templates/indexer/indexer.yaml [hunk 1]
```diff
@@ -0,0 +1,126 @@
+{{- if .Values.shardCount }}
+{{- $replicas := .Values.shardCount | int }}
```
Comment (JoanFM) [inline pqlite-helm/templates/indexer/indexer.yaml:2]: here there seems to be a confusion between `replica` and `shard`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: pqlite-helm/templates/indexer/indexer.yaml (hunk 1 above)
Comment (numb3r3) [reply pqlite-helm/templates/indexer/indexer.yaml:2]: yes, i will change the name

**Comment by numb3r3** [comment]
I will wait until https://github.com/jina-ai/jina/pull/4223

**Comment by numb3r3** [comment]
> Yes, the `gateway` is optional. The initial motivation for this PR is to support **PQLite-As-A-Service**, and then `gateway` is added to expose the HTTP (or GRPC) service.

Now, I would to like to keep `gateway` for potential features (e.g., access by Jina client directly). In practice, the recommended way is to use pqlite as **external pod** in a flow, as shown in the **Step 5**.

**Pull request merged**